### PR TITLE
PR: Single-stage optimization add-ons

### PR DESCRIPTION
**Pull request by rogeriojorge**
The pull request introduces a few features to allow single-stage optimizations to be performed using SIMSOPT. In particular, these are the minimal set of changes needed to fully reproduce the results in https://arxiv.org/abs/2302.10622.
Brief list of changes:
- Call to jac() in finite_difference allows non-dofs to be broadcasted in an MPI setting
- Range of surface to be initialized by VMEC as an optional argument (then used in from_nphi_ntheta())
- Squared flux objective now written as a surface integral of B.n/B instead of only B.n.

### CODE REVIEW

File: src/simsopt/_core/finite_difference.py [hunk 1]
```diff
@@ -209,6 +211,9 @@ def _jac(self, x: RealArray = None):
         nparams = opt.dof_size
         # Make sure all leaders have the same x0.
         mpi.comm_leaders.Bcast(x0)
+        if not args == (): non_dofs = np.array(args)
+        else: non_dofs = None
```
Comment (mbkumar) [inline src/simsopt/_core/finite_difference.py:215]: Not pythonic. Use 
`non_dofs = np.array(args) if args else None`

File: src/simsopt/objectives/least_squares.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ def unweighted_residuals(self, x=None, *args, **kwargs):
                     break
 
                 output = np.array([out]) if not np.ndim(out) else np.asarray(out)
-                output -= self.goals[i]
+                output = output - self.goals[i]
                 if self.first_eval:
```
Comment (mbkumar) [inline src/simsopt/objectives/least_squares.py:166]: Why this change?

File: src/simsopt/_core/finite_difference.py [hunk 3]
```diff
@@ -360,7 +373,7 @@ def jac(self, x: RealArray = None, *args, **kwargs):
         self.mpi.mobilize_leaders(ARB_VAL)  # Any value not equal to STOP
         self.mpi.comm_leaders.bcast(x, root=0)
 
-        jac, xs, evals = self._jac(x)
+        jac, xs, evals = self._jac(x, args)
```
Comment (mbkumar) [inline src/simsopt/_core/finite_difference.py:376]: `self.non_dofs` in place of `args`?

File: src/simsopt/_core/finite_difference.py [hunk 4]
```diff
@@ -253,6 +258,7 @@ def _jac(self, x: RealArray = None):
                 x = xs[:, j]
                 mpi.comm_groups.bcast(x, root=0)
                 opt.x = x
+                self.opt.non_dofs = non_dofs
```
Comment (mbkumar) [inline src/simsopt/_core/finite_difference.py:261]: How is the non_dofs attribute used in the optimizable object? Is there any convention or standardization you have in mind here?

File: src/simsopt/_core/finite_difference.py (hunk 1 above)
Comment (rogeriojorge) [reply src/simsopt/_core/finite_difference.py]: Fixed in commit https://github.com/hiddenSymmetries/simsopt/pull/301/commits/e68266e3a0da0392ad541225d835a9eace25122d

File: src/simsopt/_core/optimizable.py [hunk 5]
```diff
@@ -1654,6 +1654,10 @@ def __init__(self, func, *args, dof_indicators=None, **kwargs):
                         non_dofs.append(arg)
                     elif dof_indicators[i] == "dof":
                         dofs.append(arg)
+                        # if dof_indicators.count("dof") == 1:
+                        #     dofs = arg
+                        # else:
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/_core/optimizable.py]: Removed in commit https://github.com/hiddenSymmetries/simsopt/pull/301/commits/47eedc3536fc11b9dacba2ea523951d2e5fddf6e

File: src/simsopt/objectives/least_squares.py (hunk 2 above)
Comment (rogeriojorge) [reply src/simsopt/objectives/least_squares.py:166]: I've had this break on a few occasions with more general optimization metrics. I can't reproduce it with the tests but the error is usually printed as
`gcc12 error: no match for 'operator*' overload operator not working`

File: src/simsopt/_core/finite_difference.py [hunk 6]
```diff
@@ -343,6 +349,13 @@ def jac(self, x: RealArray = None, *args, **kwargs):
         ARB_VAL = 100
         logger.debug("Entering jac evaluation")
 
+        try:
+            if not args == (): non_dofs = np.array(args)
+            else: non_dofs = None
+        except Exception as e: print(e)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/_core/finite_difference.py]: Fixed in commit https://github.com/hiddenSymmetries/simsopt/pull/301/commits/e9acd7aca7ccda4b5db1ca3bdb7eb256aef91929

File: src/simsopt/_core/optimizable.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/_core/optimizable.py]: Removed in commit https://github.com/hiddenSymmetries/simsopt/pull/301/commits/47eedc3536fc11b9dacba2ea523951d2e5fddf6e

File: src/simsopt/_core/finite_difference.py (hunk 4 above)
Comment (rogeriojorge) [reply src/simsopt/_core/finite_difference.py:260]: I was thinking that it should be an array of doubles. I have no other use cases in mind.

**Comment by rogeriojorge** [comment]
For a single-stage example with dofs and non-dofs, check
https://github.com/hiddenSymmetries/simsopt/blob/c1439e4db78ac76a518813d02b13c290b6763740/examples/3_Advanced/single_stage_optimization.py

File: src/simsopt/_core/finite_difference.py (hunk 4 above)
Comment (rogeriojorge) [reply src/simsopt/_core/finite_difference.py:260]: See example here
https://github.com/hiddenSymmetries/simsopt/blob/rj/single_stage_PR/examples/3_Advanced/single_stage_optimization_finite_beta.py#L223

File: src/simsopt/objectives/fluxobjective.py [hunk 7]
```diff
@@ -29,19 +29,31 @@ class SquaredFlux(Optimizable):
           in ``phi`` and ``theta`` direction.
     """
 
-    def __init__(self, surface, field, target=None):
+    def __init__(self, surface, field, target=None, local=True):
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/objectives/fluxobjective.py:36]: Added docstring in commit 
https://github.com/hiddenSymmetries/simsopt/commit/912708029c54565b78d43ff68fbc0d58a2398e06

File: src/simsopt/objectives/fluxobjective.py [hunk 8]
```diff
@@ -29,19 +29,31 @@ class SquaredFlux(Optimizable):
...
         self.field.set_points(xyz.reshape((-1, 3)))
+        self.local = local
         Optimizable.__init__(self, x0=np.asarray([]), depends_on=[field])
 
     def J(self):
+        xyz = self.surface.gamma()
         n = self.surface.normal()
-        Bcoil = self.field.B().reshape(n.shape)
-        Btarget = self.target if self.target is not None else []
-        return sopp.integral_BdotN(Bcoil, Btarget, n)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/objectives/fluxobjective.py:44]: Done in commit https://github.com/hiddenSymmetries/simsopt/commit/7e7fafb9243b948d9fdba289b5430de946abf6ab

**Comment by rogeriojorge** [comment]
Ready to be reviewed again

File: src/simsopt/_core/optimizable.py [hunk 9]
```diff
@@ -93,6 +93,8 @@ def __init__(self,
         """
         if x is None:
             x = np.array([])
+        elif np.isscalar(x):
```
Comment (rogeriojorge) [inline src/simsopt/_core/optimizable.py:96]: If this is not added the tests don't pass

**Comment by rogeriojorge** [comment]
I am not really sure why the tests are not passing, @landreman @mbkumar could you take a look?

**Comment by rogeriojorge** [comment]
Now ready to be reviewed. All tests passed and both a vacuum and finite beta case were added as examples.

File: src/simsopt/_core/finite_difference.py (hunk 4 above)
Comment (landreman) [reply src/simsopt/_core/finite_difference.py:260]: This smells funny to me, because `non_dofs` is an attribute only of objects created by `make_optimizable()`. Most optimizable objects don't have a `non_dofs` attribute.

**Review by landreman — COMMENTED** [review]
If I understand correctly, the edits to `finite_difference.py` are made because in `single_stage_optimization_finite_beta.py`, you apply `MPIFiniteDifference` to the result of `make_optimizable()` on `fun_J`, and you only want to apply finite differencing to the surface dofs and not to the coil dofs. I don't think anything in `single_stage_optimization.py` requires those changes to `finite_difference.py`. Is this correct? 

The present implementation uses `non_dofs`, which is only an attribute of objects created by `make_optimizable()`, not by other Optimizable objects. It would be nice to keep the finite difference machinery written in a way that applies to any Optimizable object, not just `make_optimizable()` ones, so let's think about if there is a way to refactor. It feels like there should be a solution in which `JF.x` is mpi-bcast-ed somewhere in the driver script, but I haven't thought through all the details.

File: examples/3_Advanced/single_stage_optimization.py
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/3_Advanced/single_stage_optimization.py:1]: Done in the latest commits

File: src/simsopt/_core/optimizable.py [hunk 10]
```diff
@@ -1704,7 +1709,10 @@ def J(self):
                     args.append(self.parents[opt_ind])
                     opt_ind += 1
                 elif self.dof_indicators[i] == 'dof':
-                    args.append(dofs[dof_ind])
+                    if self.dof_indicators.count("dof") == 1:
+                        args.append(dofs)
+                    else:
+                        args.append(dofs[dof_ind])
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply src/simsopt/_core/optimizable.py:1715]: This was implemented in order to allow dofs to be a 1D array put by the user in the optimizable object. To make it work with both 1 dof and several dofs cases, I changed the line

`dofs = self.local_full_x`

to

`dofs = np.atleast_2d(self.local_full_x)`

This should not affect previous runs, only the new ones where dofs would be a 1D array.

File: examples/3_Advanced/single_stage_optimization_finite_beta.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+J_ALS = ARCLENGTH_WEIGHT * sum(Jals)
+J_LENGTH_PENALTY = LENGTH_CON_WEIGHT * sum([QuadraticPenalty(Jls[i], LENGTH_THRESHOLD) for i in range(len(base_curves))])
+JF = Jf + J_CC + J_LENGTH + J_LENGTH_PENALTY + J_CURVATURE + J_MSC
+##########################################################################################
+pprint(f'  Starting optimization')
+##########################################################################################
+# Initial stage 2 optimization
+##########################################################################################
+
+
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/3_Advanced/single_stage_optimization_finite_beta.py:137]: Explained in the latest commits.

File: examples/3_Advanced/single_stage_optimization_finite_beta.py [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+    J_stage_1 = prob.objective()
+    dofs_coils = np.ravel(dofs_coils)
+    if np.sum(JF.x != dofs_coils) > 0:
+        JF.x = dofs_coils
+    if run_vcasing:
+        try:
+            vc = VirtualCasing.from_vmec(vmec, src_nphi=vc_src_nphi, trgt_nphi=nphi_VMEC, trgt_ntheta=ntheta_VMEC)
+            Jf = SquaredFlux(surf, bs, local=True, target=vc.B_external_normal)
+            if np.sum(Jf.x != dofs_coils) > 0: Jf.x = dofs_coils
+            JF.opts[0].opts[0].opts[0].opts[0].opts[0] = Jf
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/3_Advanced/single_stage_optimization_finite_beta.py:178]: Simplified the code by adding the change

`Jf.target = vc.B_external_normal`

File: examples/3_Advanced/single_stage_optimization_finite_beta.py [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+            if np.sum(JF.x != dofs_coils) > 0: JF.x = dofs_coils
+        except Exception as e:
+            J = JACOBIAN_THRESHOLD
+            Jf = JF.opts[0].opts[0].opts[0].opts[0].opts[0]
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/3_Advanced/single_stage_optimization_finite_beta.py:182]: Yes, done

File: examples/3_Advanced/single_stage_optimization_finite_beta.py [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    if np.sum(JF.x != dofs_coils) > 0:
+        JF.x = dofs_coils
+    if run_vcasing:
+        try:
+            vc = VirtualCasing.from_vmec(vmec, src_nphi=vc_src_nphi, trgt_nphi=nphi_VMEC, trgt_ntheta=ntheta_VMEC)
+            Jf = SquaredFlux(surf, bs, local=True, target=vc.B_external_normal)
+            if np.sum(Jf.x != dofs_coils) > 0: Jf.x = dofs_coils
+            JF.opts[0].opts[0].opts[0].opts[0].opts[0] = Jf
+            if np.sum(JF.x != dofs_coils) > 0: JF.x = dofs_coils
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/3_Advanced/single_stage_optimization_finite_beta.py:180]: Now catching only OjectiveFailure here

File: examples/3_Advanced/single_stage_optimization_finite_beta.py [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+    grad = np.concatenate((grad_with_respect_to_coils, grad_with_respect_to_surface))
+
+    # Remove spurious files
+    for vcasing_file in glob.glob("vcasing*"): os.remove(vcasing_file)
+    for jac_file in glob.glob("jac_log_*"): os.remove(jac_file)
+    os.chdir(parent_path)
+    for vcasing_file in glob.glob("vcasing*"): os.remove(vcasing_file)
+    for jac_file in glob.glob("jac_log_*"): os.remove(jac_file)
+    os.chdir(this_path)
+    for vcasing_file in glob.glob("vcasing*"): os.remove(vcasing_file)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/3_Advanced/single_stage_optimization_finite_beta.py:217]: Added filename=None and not removing any 'spurious files'.

File: src/simsopt/_core/optimizable.py (hunk 9 above)
Comment (rogeriojorge) [reply src/simsopt/_core/optimizable.py:96]: This is now not needed as dofs are now handled correctly in J

**Comment by rogeriojorge** [comment]
Updated branch with latest SIMSOPT master branch

**Review by landreman — APPROVED** [review]
